### ISSUE TITLE
Context manager generator yielding in exception handler, breaks re-raise of exception from with statements

### ISSUE THREAD

**Issue by CodeisCold** [Nuitka/Nuitka#2811]
python version: 3.9.18
nuitka version: 2.1.5
`pip list` result:
```
Package           Version
----------------- -------
asgiref           3.8.1
Django            4.2.11
Nuitka            2.1.5
ordered-set       4.1.0
pip               23.0.1
setuptools        58.1.0
sqlparse          0.4.4
typing_extensions 4.11.0
zstandard         0.22.0

[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: pip install --upgrade pip
```

code example:
```python
from asgiref.local import Local


if __name__ == '__main__':
    a = Local(thread_critical=True)
    getattr(a, 'default')
```

nuitka package cmd: `python -m nuitka --onefile --show-progress test.py`

expected result:
```
AttributeError: '_thread._local' object has no attribute 'default'
```

what i get:
```
Traceback (most recent call last):
  File "/tmp/onefile_81632_1712738386_558915/test.py", line 6, in <module>
    getattr(a, 'default')
  File "/tmp/onefile_81632_1712738386_558915/asgiref/local.py", line 118, in __getattr__
RuntimeError: No active exception to reraise
```

**Comment by kayhayen** [comment]
This is looking like an issue of compatibility with `getattr` on the storage while an exception was active in the context generator function used here:

```
    def __getattr__(self, key):
        with self._lock_storage() as storage:
            return getattr(storage, key)
```

**Comment by kayhayen** [comment]
With Python 3.10, this is not happening.

**Comment by kayhayen** [comment]
I believe, we finally need to generate the getattr, hasattr, variations that are floating around, where we also can have a look at taking advantage of the ability to cache attribute lookups starting with 3.11, this probably a failure to save and restore an exception, dropping the failure of `getattr` attempt on some object method. It seems this one is confusing it:

```
    def __getattr__(self, key):
        storage_object = self._data.get({})
        try:
            return storage_object[key]
        except KeyError:
            raise AttributeError(f"{self!r} object has no attribute {key!r}")
```

**Comment by kayhayen** [comment]
I am confused to see that you report using 3.9, but alas, maybe there is that too, not sure now. You were not using the proper reporting output to be sure what it really was.

**Comment by CodeisCold** [comment]
```
2.1.5
Commercial: None
Python: 3.9.18 (main, Mar 13 2024, 11:01:03) 
Flavor: Unknown
Executable: /root/pycharm_project_593/web/venv/bin/python
OS: Linux
Arch: aarch64
Distribution: Kylin V10
Version C compiler: /usr/bin/gcc (gcc 7.3.0).
```

**Comment by kayhayen** [comment]
It's strange that my 3.10 didn't do it, but totally possible the versions of `asgiref` just didn't align, and then this error would be more general indeed. Once I get to it, I will check it out.

**Comment by sreid19** [comment]
Wanted to jump in here - I am also getting this issue with Django and asgiref:

2.3.11
Commercial: None
Python: 3.10.0 | packaged by conda-forge | (default, Nov 10 2021, 13:20:59) [MSC v.1916 64 bit (AMD64)]
Flavor: Anaconda Python
Executable: C:\Users\scott\anaconda3\envs\exe\python.exe
OS: Windows
Arch: x86_64
WindowsRelease: 11
Version C compiler: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\bin\Hostx64\x64\cl.exe (cl 14.3).

asgiref==3.8.1
Django==5.0.6

Traceback (most recent call last):
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\startup.py", line 38, in <module>
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\__init__.py", line 24, in setup
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\apps\registry.py", line 116, in populate
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\apps\config.py", line 269, in import_models
  File "importlib.py", line 126, in import_module
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\contrib\auth\models.py", line 3, in <module django.contrib.auth.models>
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\contrib\auth\base_user.py", line 59, in <module django.contrib.auth.base_user>
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\db\models\base.py", line 143, in __new__
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\db\models\base.py", line 371, in add_to_class
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\db\models\options.py", line 243, in contribute_to_class
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\utils\connection.py", line 15, in __getattr__
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\django\utils\connection.py", line 58, in __getitem__
  File "C:\Users\scott\website\exe_only\site\STARTU~1.DIS\asgiref\local.py", line 117, in __getattr__
RuntimeError: No active exception to reraise

Command used:
python -m nuitka startup.py --include-data-files=params.txt=params.txt --enable-plugin=tk-inter --standalone --include-package=django --include-package-data=django --include-package=django.views --include-package-data=django.views

To reproduce

import django
from django.conf import settings
import os

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "settings")

settings.configure(INSTALLED_APPS=[
   "django.contrib.admin",
   "django.contrib.auth",
   "django.contrib.contenttypes",
   "django.contrib.sessions",
   "django.contrib.messages",
   "django.contrib.staticfiles",
])
django.setup()

**Comment by kayhayen** [comment]
There has been some Django progress recently, but on an empty Django project with 3.12, I am now seeing this for myself easily, and will be able to reproduce and fix this hopefully. I noticed that this is in a spot, where I recently added an assertion, because of issues seen in other places, hopefully it turns out to be triggering with `--debug` in my current development version too.

**Comment by kayhayen** [comment]
So, this seems to be about the with statement, and either its re-formulation or try/finally in generally, being able to become confused. It seems generated code doesn't fetch the exception when checking it in the exception handler, while it ran code to change things there, so the re-raise is triggered, although it ought not to, since there wasn't an actual exception.

This can be indicative of 2 things, the re-raise is checking for the wrong value, and/or the clearing in getattr sets the "wrong" value, and there is maybe a need to re-factor the exception handling code to fetch the exception so it cannot get changed behind its back.

**Comment by kayhayen** [comment]
I managed to reproduce this one with a smaller program.

```
import asyncio
import contextlib
import contextvars


class _CVar:
    def __init__(self) -> None:
        self._data = contextvars.ContextVar("mini")

    def __getattr__(self, key):
        storage_object = self._data.get({})
        try:
            return storage_object[key]
        except KeyError:
            raise AttributeError(f"{self!r} object has no attribute {key!r}")

    def __setattr__(self, key, value):
        if key == "_data":
            return super().__setattr__(key, value)

        storage_object = self._data.get({})
        storage_object[key] = value
        self._data.set(storage_object)


key = "default"


@contextlib.contextmanager
def _lock_storage():
    _storage = _CVar()
    try:
        # Will raise RuntimeError.
        asyncio.get_running_loop()
    except RuntimeError:
        yield _storage


def f():
    with _lock_storage() as storage:
        x = getattr(storage, key)


f()
```

The error occurs only when the generator yield inside the exception handler for the `RuntimeError` provoked. For the duration of the exception handler, the state of the published exception is put out and removed afterwards. This apparently cleared the exception that the re-formulation of the with statement wants to raise, and has published in its exception handler, where it called `__exit__` to check it.

So something is incompatible or buggy here. Either the generator is wrong to clear something there, or the with statement re-formulation ought to not publish the exception to the `__exit__` call and raise it from published. Quite likely, the contextmanager function closes the generator inside of it, and that changes the published exception state, and probably must not happen, so this could well be the issue here, which would be good, since then publishing it is only overhead maybe, but it's too early to call.

I will make sure that `_CVar` has no role in this, and also I noticed a segfault crash if this is not happening in a function, but on the module level, which is also very confusing and definitely ought to not happen, but might be indicative of a corruption happening with generator and publish statements interactions not being bug free.

**Comment by kayhayen** [comment]
So, I got this down to this, and the segfault was probably just debugging code of mine, so false alarm there:

```
import contextlib


def get_running_loop():
    raise RuntimeError


class _CVar:
    def __getattr__(self, key):
        raise AttributeError("object has no attribute %s" % key)


key = "default"


@contextlib.contextmanager
def _lock_storage():
    _storage = _CVar()
    try:
        # Will raise RuntimeError.
        get_running_loop()
    except RuntimeError:
        yield _storage


with _lock_storage() as storage:
    x = getattr(storage, key)
```

This shows exactly the two features interacting badly, no contextvars or asyncio involved anymore. With Python2 this doesn't happen, since publishing of exceptions to handlers was not done there (yet). But I tried Python3 down to 3.10, and all are the same. I somehow initially suspected version dependence and `getattr` but that was all misguided.

**Comment by kayhayen** [comment]
Also, checking our test for `with` statements, it covers the publication of the exception in `sys.exc_info()` that should be done indeed, so most likely, the closing of the generator is failing to not unpublish the wrong exception, one it didn't put there or something like that.

**Comment by kayhayen** [comment]
So, I figured it out, when switching to and from a coroutine, we need to potentially save and restore an exception the generator handles, which overwrote the public exception. So the generator yielding in an exception handler to a caller that itself is in an exception handler, creates a conflict where the exception handler removes the exception made public (i.e. visible to `sys.exc_info()` that is not the own it owns.

This has to be solved with saving and restoring the generator frame exception before enter/leave of a yield (if any), abd should be very doable. I am not entirely sure, but I hope to have something for a hotfix during this week. 

This is kind of a surprising compatbility issue, given the incredible amount of code that was never affected by it. But you never stop learning. And it affects all of Python3, Python2 didn't have that those frame local exceptions interactions with the currently public exception.

**Comment by kayhayen** [comment]
So, I did actually work on this a bunch. I found that Nuitka actually implements the save/restore of public exceptions at being and end of the exception handlers in generators. I also found that starting with 3.7, the internal structures allow for the published exception to be a stack, where on entry the generator could push its own empty state and work on only that, and remove it when leaving the generator for yield.

In best experimental computer science fashion, I added that stack and changed code over to work on that stack type internally, only to find it doesn't happen. I believe I mixed up optimization and correction on the way. The bug is applicable to Python3.6 after all, and doing something more optimal wasn't going to do more than hiding it probably. Right now, the stack is supposed to be implemented already. And I think, it needs to do the publication and local restore of what it was before in the generator, as was as save and restore when entering and yielding and returning from that.

So, the plan is now to work with traces until I figure out, where the wrong thing actually happens yield now. And to make small scripts to examine, if the exception from the generator leaks or if I can corrupt it with anything but said with statement, which appears to be a bit complex to begin with. 

So far this fails:

```
exec("""

class _GeneratorContextManager:
    def __init__(self, func, args, kwds):
        self.gen = func(*args, **kwds)
        self.func, self.args, self.kwds = func, args, kwds

    def __enter__(self):
        # do not keep args and kwds alive unnecessarily
        # they are only needed for recreation, which is not possible anymore
        del self.args, self.kwds, self.func
        try:
            return next(self.gen)
        except StopIteration:
            raise RuntimeError("generator didn't yield") from None

    def __exit__(self, typ, value, traceback):
        print("__exit__", "args", typ, value, traceback, "exc", sys.exc_info())

        assert typ is not None
        try:
            self.gen.throw(typ, value, traceback)
        except BaseException as exc:
            print("exception from generator throw", exc, exc is value, sys.exc_info())
            assert exc is value

            print("returning false", exc, sys.exc_info())
            return False
        raise RuntimeError("generator didn't stop after throw()")

    def __call__(self, func):
        def inner(*args, **kwds):
            with self:
                return func(*args, **kwds)

        return inner


def contextmanager(func):
    def helper(*args, **kwds):
        return _GeneratorContextManager(func, args, kwds)

    return helper

def get_running_loop():
    raise RuntimeError


class _CVar:
    def __getattr__(self, key):
        raise AttributeError("object has no attribute %s" % key)
""")

@contextmanager
def _lock_storage():
    _storage = _CVar()
    try:
        # Will raise RuntimeError.
        get_running_loop()
    except RuntimeError:
        yield _storage


def f():
    with _lock_storage() as storage:
        x = getattr(storage, key)
```

I avoid compiled code as much as possible, and I have a reduced contextmanager in that exec, and hopefully that will be possible to isolate the issue much more. Somehow it ought to be possible to get at what it really is.

And from there, for Python3.7+ I should then be able to recognize what parts of the code can be replaced with the exception stack they now use internally. This one is unfortunately not as easy to find as I hoped for, but I mean to explore it this weekend until I have it.

**Comment by kayhayen** [comment]
So, I enhanced the tracing, and I found something very suspicious happening now if I trace and display the exception changes:

```
Execute: MiniWithGeneratorExceptionPublish.py:66 <Node 'STATEMENT_RERAISE_EXCEPTION'>
RESTORE_ERROR_OCCURRED:
current_exc=<class 'AttributeError'> refcnt 12 |AttributeError('object has no attribute default') refcnt 4  <- context <NULL>|<traceback object at 0x7f6accdc3d40> refcnt 1 
exception from generator throw object has no attribute default True (<class 'AttributeError'>, AttributeError('object has no attribute default'), <traceback object at 0x7f6accba6980>)
returning false object has no attribute default (<class 'AttributeError'>, AttributeError('object has no attribute default'), <traceback object at 0x7f6accba6980>)
Execute: MiniWithGeneratorExceptionPublish.py:76 <Node 'STATEMENT_RERAISE_EXCEPTION'>
Execute: MiniWithGeneratorExceptionPublish.py:76 <Node 'STATEMENT_RESTORE_FRAME_EXCEPTION' with {'preserver_id': 1}>
SET_CURRENT_EXCEPTION:
thread_exc=<class 'NoneType'> refcnt 9 |None refcnt 4789 |<NULL>
Execute: MiniWithGeneratorExceptionPublish.py:76 <Node 'STATEMENT_RERAISE_EXCEPTION'>
Execute: MiniWithGeneratorExceptionPublish.py:76 <Node 'STATEMENT_CONDITIONAL'>
Execute: MiniWithGeneratorExceptionPublish.py:76 <Node 'STATEMENT_RERAISE_EXCEPTION'>
```

This is missing some traces, but I find it kind of suspcious, how there are two re-raises one after another, and that after the re-raise (which is line wise, the with statement and its re-formulation) there is a restore of the frame exeption, which there was none (it does the SET_CURRENT_EXCEPTION) and then a re-raise of that. This might be different between current and published exception, and be just fine, but the next 2 re-raises, I don't see those come to pass.

**Comment by kayhayen** [comment]
Makes me really wish, that for nodes, we would have annotations, potentially, such that they are recognizable. But I also would hate if that ends up needing per-node memory in the instance dictionary, but seriously this would be a much better trace if the intent of these re-raises was visible. Maybe I can add something for debug mode only.

**Comment by kayhayen** [comment]
I think, looking at the code generation, something that's going on, is that exceptions are preserved with an id, the preserved it, from there they are restored, and then sometimes `STATEMENT_RERAISE_EXCEPTION` is supposed to find it there, by virtue of this code telling it where to get it from:

```
    (
        keeper_exception_state_name,
        _keeper_lineno,
    ) = context.getExceptionKeeperVariables()
```

That's done, despite there being absolutely no doubt, which one this is supposed to address, however, it sort of could be the case, that inner ones, could overwrite it. Basically this about shuffling the kept exception state into the one that re-represents the current one, and what now mean to do, is to add a specialized form of re-raise that knows it is to fetch from a preserver id, and then without changing code generation, to assert at first, that's it is the correct one.

**Comment by kayhayen** [comment]
So, I managed to isolate it. I found that Python3 needs code for the `yield` that Nuitka currently doesn't have. When it enters the try block, the generator doesn't have any exception set (that us during `__enter__`, it stores that, and then later, when the generator is called again for `__exit__`, it restores that. My previous attempts to observe this in Python code failed, I had to dive deep into the C code, to so it happen, when the uncompiled (and reduced) context manager throws the `AttributeError` that is currently published into it.

I found that in 3.6 (to which Nuitka's implementation is most similar right now) there is code like this:

```
fast_yield:
    if (co->co_flags & (CO_GENERATOR | CO_COROUTINE | CO_ASYNC_GENERATOR)) {

        /* The purpose of this block is to put aside the generator's exception
           state and restore that of the calling frame. If the current
           exception state is from the caller, we clear the exception values
           on the generator frame, so they are not swapped back in latter. The
           origin of the current exception state is determined by checking for
           except handler blocks, which we must be in iff a new exception
           state came into existence in this frame. (An uncaught exception
           would have why == WHY_EXCEPTION, and we wouldn't be here). */
        int i;
        for (i = 0; i < f->f_iblock; i++) {
            if (f->f_blockstack[i].b_type == EXCEPT_HANDLER) {
                break;
            }
        }
        if (i == f->f_iblock)
            /* We did not create this exception. */
            restore_and_clear_exc_state(tstate, f);
        else
            swap_exc_state(tstate, f);
    }
```

So, basically, when "sending" values into the generator, we need to keep track of the published exception in the generator object, to know what to restore.

I think a yielding generator is supposed to see the published exception, but it's not allowed to overwrite it, but instead, it must save it until it resumes again in a protected space, and so things become very similar to this code from Python 3.6 and ought to work correctly. I will be able to try this out now and report back if that manages to pass tests.

**Comment by kayhayen** [comment]
So, saving the exception at entry, and if it's not None, makes the problem go away and passes basic tests apparently, so this is looking good. I have only made the code change for 3.11+ so far, but older versions will be no issue, it's just that the published exception changed format from tripple `(type, value, tb)`  to normalized value in that version.

I am currently running the CPython 3.11 test suite, to see if it's finding anything of relevance, since this could possibly be a very invasive change. I think there might be an issue, that on next yield, the replaced published exception is not restored, but I am just assuming, this is just a matter of doing a bit more work, namely to keep it around and consider restoring it for re-entry of the generator.

**Comment by kayhayen** [comment]
So, good news, test suite ran successfully. It was actually not to be expected to see any changes, since obviously the test suite previously didn't cover this behavior. I am going to complete the change for all versions and also coroutines and asyncgen, which should have the same issues, and it will be in the next hotfix or the one after it hopefully

**Comment by kayhayen** [comment]
The fix for 3.11 and only generators is available. More coverage will be added before it goes to the hotfix, but this can be tested already.

Thanks for your report. This is fixed on the factory, which is a development version that is under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 
	
Feedback on whether this is working is very welcome. Please do not share plans to do it; only confirm or deny that it is working.

**Comment by kayhayen** [comment]
The underlying issue has been solved and is part of the stable release, starting in some 2.6 hotfix, but definitely for 2.7
